src/zarr/storage/_cache.py

```diff
@@ -0,0 +1,735 @@
+        store : Store
+            The underlying store to wrap with caching.
+        max_size : int
+            The maximum size that the cache may grow to, in number of bytes.
+        **kwargs : Any
+            Additional keyword arguments passed to the store constructor.
+
+        Returns
+        -------
+        LRUStoreCache
+            The opened cache store instance.
+        """
+        cache = cls(store, max_size=max_size, **kwargs)
+        await cache._open()
+        return cache
+
+    def with_read_only(self, read_only: bool = False) -> "LRUStoreCache":
+        """
+        Return a new LRUStoreCache with a new read_only setting.
+
+        Parameters
+        ----------
+        read_only
+            If True, the store will be created in read-only mode. Defaults to False.
+
+        Returns
+        -------
+        LRUStoreCache
+            A new LRUStoreCache with the specified read_only setting.
+        """
+        # Create a new underlying store with the new read_only setting
+        underlying_store = self._store.with_read_only(read_only)
+        return LRUStoreCache(underlying_store, max_size=self._max_size, read_only=read_only)
+
+    def _normalize_key(self, key: Any) -> str:
+        """Convert key to string if it's a Path object, otherwise return as-is"""
+        if isinstance(key, Path):
+            return str(key)
+        return str(key)
+
+    def __getstate__(
+        self,
+    ) -> tuple[
+        Store,
+        int,
+        int,
+        list[str] | None,
+        dict[Any, Any],
+        dict[str | None, list[str]],
+        OrderedDict[str, Any],
+        int,
+        int,
+        bool,
+        bool,
+    ]:
+        return (
+            self._store,
+            self._max_size,
+            self._current_size,
+            self._keys_cache,
+            self._contains_cache,
+            self._listdir_cache,
+            self._values_cache,
+            self.hits,
+            self.misses,
+            self._read_only,
+            self._is_open,
+        )
+
+    def __setstate__(
+        self,
+        state: tuple[
+            Store,
+            int,
+            int,
+            list[str] | None,
+            dict[Any, Any],
+            dict[str | None, list[str]],
+            OrderedDict[str, Any],
+            int,
+            int,
+            bool,
+            bool,
+        ],
+    ) -> None:
+        (
+            self._store,
+            self._max_size,
+            self._current_size,
+            self._keys_cache,
+            self._contains_cache,
+            self._listdir_cache,
+            self._values_cache,
+            self.hits,
+            self.misses,
+            self._read_only,
+            self._is_open,
+        ) = state
+        self._mutex = Lock()
+
+    def __len__(self) -> int:
+        return len(self._keys())
+
+    def __iter__(self) -> Iterator[str]:
+        return self.keys()
+
+    def __contains__(self, key: Any) -> bool:
```

Comment:
```suggestion
    def __contains__(self, key: str | Path) -> bool:
```
I think this should have tighter typing?